Estebanreyl/dataplane runtime sdk v2 prerelease

estebanreyl opened the pull request:
This preview release  contains the fully functional runtime sdk ready to be packaged as a nuget for the ACR runtime API. It was built from the recently approved  https://github.com/Azure/azure-rest-api-specs/pulls?utf8=%E2%9C%93&q=is%3Apr+estebanreyl

This release:
- Adds support for most of our Blob end point API's
- Adds support for multiple Image manifest types
- Adds oauth2 support (since v1 was overtaken by v2 before it could be approved)
- Includes many fixes as described by the last two swagger releases

**Renames many functions as described below:**
GetManifestAttributesAsync: Manifests.GetAttributesAsync
GetManifestAsync: Manifests.GetAsync
UpdateManifestAttributesAsync: Manifests.UpdateAttributesAsync
CreateManifestAsync: Manifests.CreateAsync
DeleteManifestAsync: Manifests.DeleteAsync
GetManifestListAsync: Manifests.GetListAsync

GetTagAttributesAsync: Tag.GetAttributesAsync
GetTagListAsync: Tag.GetListAsync
UpdateTagAttributesAsync: Tag.UpdateAttributesAsync

GetRepositoryListAsync: Repository.GetListAsync
GetRepositoryAttributesAsync : Repository.GetAttributesAsync
DeleteRepositoryAsync : Repository.DeleteAsync
UpdateRepositoryAttributesAsync : Repository.UpdateAttributesAsync

**New functions include:**
Blob.GetAsync
Blob.CheckAsync
Blob.StartUploadAsync
Blob.CancelUploadAsync
Blob.GetStatusAsync
Blob.GetChunkAsync
Blob.CheckChunkAsync
Blob.MountAsync
Blob.StartUploadAsync
Blob.UploadAsync
Blob.EndUploadAsync
Blob.DeleteAsync

RefreshTokens.GetFromExchangeAsync
AccessTokens.GetAsync
AccessTokens.GetFromLoginAsync
V2Support.CheckAsync

New ease of use constructor and acrClientCredentials also added.

Closes: Azure/azure-rest-api-specs#6773

Commit:
Add v2 SDK into clean branch because of absurd number of merge and merge conflicts before.

Commit:
Added missing config file from previous commit

adewaleo commented:
@AlexGhiondea, @erich-wang, a review would be appreciated.

AlexGhiondea commented:
Is this SDK autogenerated?

amishra-dev reviewed (changes requested):
can you change the name of classes and namespaces that have Blob on it to DockerBlob. I do not want this package to conflict with Azure Storage blob.

adewaleo commented:
@AlexGhiondea, I believe it is auto generated with some modifications.

@amishra-dev I will look into this. We could rename it **RegistryBlob** or **ACRBlob**

@estebanreyl can you confirm?

estebanreyl commented:
The blob naming comes directly from the swagger generation. We could update that and then eventually have it go through but at the end of the day our blob paths would be under the container registry client and should thus be unambiguous especially since their namespace is distinct from that of Azure Storage blob which in C# being a strongly typed language should make it hard to make the mistake since blob by itself for us is just a method disambiguation level. Additionally, renaming them to ACRBlob or DockerBlob are not very representative as the blobs in question aren't either unique to ACR, nor docker. In fact blobs in our case just represent arbitrary data streams.

adewaleo commented:
> @adewaleo , @amishra-dev The blob naming comes directly from the swagger generation. We could update that and then eventually have it go through but at the end of the day our blob paths would be under the container registry client and should thus be unambiguous especially since their namespace is distinct from that of Azure Storage blob which in C# being a strongly typed language should make it hard to make the mistake since blob by itself for us is just a method disambiguation level its not even an object type but an instance. Additionally, renaming them to ACRBlob or DockerBlob are not very representative as the blobs in question aren't either unique to ACR, nor docker. In fact blobs in our case just represent arbitrary data streams.

This makes sense @estebanreyl. I think @amishra-dev 's fear is that someone could be using both namespaces (Microsoft.Azure.ContainerRegistry and [Azure.Storage.Blobs](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs)) in the same program. i.e methods to work with azure storage account blobs and azure container registry blobs, both of which are binary streams of data.

But as @estebanreyl  has said, these are auto generated from our swagger and as C# is strongly typed users are unlikely to have issues mixing the two.

adewaleo commented:
@amishra-dev , with the recent mail thread do you still think any classes / methods still need to be remained? Could you give examples where you feel the current naming could be unclear?

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+                throw new Exception("AAD token authorization requires you to provide the AAD_access_token");
+            }
+            // Proofing in case passed in loginurl includes https start.
+            if (loginUrl.StartsWith("https://"))
```
Consider casing here

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+            }
+            else
+            {
+                string operation = "https://" + _loginUrl + request.RequestUri.AbsolutePath;
```
Consider using string interpolation instead of string concatenation.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+        /// the oauth2 endpoint improving efficiency.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+        public string getAcrAccessToken(string scope)
```
Public method should start with upper case letter.

AlexGhiondea replied:
Or should this become private?

adewaleo replied:
Seems like these are helpers and should be private. I will change the access modifier and convert to pascal case.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+
+        public async Task<string> getScope(string operation, string method, string path)
```
Upper case for method name.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+
+        public async Task<string> getScope(string operation, string method, string path)
+        {
+
```
Extra line. There are several places in the file with inconsistent spacing as well.

adewaleo replied:
Are you referring to the extra line before the code. Is this a style issue?

AlexGhiondea replied:
Yeah.. We are trying to standardize to a set of style conventions in the repo -- that being said, I don't feel strongly about it :).

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+        public async Task<string> getScope(string operation, string method, string path)
+        {
+
+            if (_acrScopes.ContainsKey(method + ">" + operation))
```
Suggestion:

Consider changing this to:

```csharp
string methodAndOperation = $"{method}>{operation}";
if (_arcScopes.TryGetValue(methodAndOperation, out string result))
{
   return result;
}
```

That reduces the string allocations as well as reduces the calls into the Dictionary methods.

adewaleo replied:
Thanks! Will do

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+            string scope;
+            try
+            {
+                response = await runtimeClient.SendAsync(new HttpRequestMessage(new HttpMethod(method), operation));
```
Should this check for the expected result?

adewaleo replied:
What do you mean by check for expected result @AlexGhiondea? Check the status code or returned response?

Here the request is made as a challenge to get the appropriate scope for this endpoint from the header.  I believe this is why we don't check the response value / method.

AlexGhiondea replied:
If you don't use the response, consider replacing it with a discardable variable named `_` (see https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#discards)

adewaleo replied:
@AlexGhiondea we actually use the variable in the next line. response.headers

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+            {
+                response = await runtimeClient.SendAsync(new HttpRequestMessage(new HttpMethod(method), operation));
+                Dictionary<string, string> data = parseHeader(response.Headers.GetValues("Www-Authenticate").FirstOrDefault());
+                scope = data.ContainsKey("scope") ? data["scope"] : hardcodedScopes(path);
```
Consider a similar approach for retrieving the value for `"scope"` from the dictionary.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+            foreach (Match part in parts)
+            {
+                string[] keyValues = part.ToString().Split('=');
+                parsed.Add(keyValues[0], headerTrim(keyValues[1]));
```
Can `keyValues` have less than 2 elements?

adewaleo replied:
It shouldn't be possible. The regex requires that the string have an '=' in it. And the array returned by String.Split('=') will have a second element in it.

That being said I personally tend to lean away from regexes, so I might change this method to not use a regex.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+        /// </summary>
+        private string headerTrim(string toTrim)
+        {
+            toTrim = toTrim.Trim();
```
This might throw for something like `"` (single quote)

adewaleo replied:
I am not sure I understand, I don't think Trim() throws errors.

Or are you talking about the calls to substring?

AlexGhiondea replied:
I meant the call to substring, but now that I look at it again, it won't be an issue since the `toTrim.Length` will at worst be 0 and the `Substring` method won't do anything.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <summary>
+        /// Provides cleanup in case Cache is getting large. 
+        ///</summary>
+        public void clearCache()
```
Public method with upper case character.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -0,0 +1,309 @@
+            _acrScopes.Clear();
+        }
+
+        static public string EncodeTo64(string toEncode)
```
Does this need to be public?

adewaleo replied:
no it doesn't I wil change this and other helpers to private

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -0,0 +1,175 @@
+            await base.ProcessHttpRequestAsync(request, cancellationToken);
+        }
+
+        static public string EncodeTo64(string toEncode)
```
This code is duplicated (see above). Can this be moved to an internal helper class instead?

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -0,0 +1,175 @@
+        private static readonly JwtSecurityTokenHandler JwtSecurityClient = new JwtSecurityTokenHandler();
+
+        // Constant to refresh tokens slightly before they are to expire guarding against possible latency related crashes
+        private TimeSpan LATENCY_SAFETY { get; set; } = TimeSpan.FromMinutes(2);
```
As this isn't a constant, I would rename this to LatencySafety. 
If this can't be changed, I would also make the field `readonly`

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -0,0 +1,175 @@
+
+        public string Value { get; set; }
+        public DateTime Expiration { get; set; }
+        public acquireCallback RefreshFn { get; set; }
```
Does this need to be public?
Does it need to be read/write?

adewaleo replied:
I don't believe it needs to be public or read/write. Will make it private and readonly

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -0,0 +1,175 @@
+
+        public string Value { get; set; }
+        public DateTime Expiration { get; set; }
+        public acquireCallback RefreshFn { get; set; }
```
I would rename this to `RefreshFunction` -- see other questions above.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -0,0 +1,175 @@
+
+        public bool NeedsRefresh()
+        {
+            return Expiration < DateTime.UtcNow.Add(LATENCY_SAFETY);
```
Question: Are both dates represented using the same date format?

adewaleo replied:
Yes, they should be. They are both DateTime objects and they should both be utc based. I can try to verify whether the jwt token's time is utc based.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AzureContainerRegistryClientExtended.cs:
```diff
@@ -0,0 +1,83 @@
+    public partial class AzureContainerRegistryClient : ServiceClient<AzureContainerRegistryClient>, IAzureContainerRegistryClient, IAzureClient
+    {
+
+        // MANUALLY ADDED FOR INTERNAL TEST PURPOSES
```
Which part was added for testing? Is that still required?

estebanreyl replied:
Oh that one is a tad confusing. Essentially the test framework requires one of the constructors there (internal AzureContainerRegistryClient(System.Uri loginUri, ServiceClientCredentials credentials, params DelegatingHandler[] handlers)) in order to run at all but autorest wasnt generating it (pretty sure it never did) so I added it so that the tests can be run. We thus do need to keep it but the comment is misleading. Should probably read "For Testing Framework use only"

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AzureContainerRegistryClientExtended.cs:
```diff
@@ -0,0 +1,83 @@
+        public AzureContainerRegistryClient(string loginUrl, ServiceClientCredentials credentials, params DelegatingHandler[] handlers) : this(handlers)
+        {
+            //Removes issues with clients potentially setting an incorrect / at the end
+            if (loginUrl.EndsWith("/")) {
```
What would happen if `loginUrl` is this string: `/` ?

adewaleo replied:
Tested, this should be fine, we just end up with an empty string if the user passes only `/`

AlexGhiondea replied:
Ok!

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -53,73 +56,63 @@ public enum LoginMode
             }
 
-            _loginUrl = loginUrl;
+            _acrScopes = new Dictionary<string, string>();
```
For the `Dictionary` I recommend being explicit about the comparer you need (ie. should it be case sensitive or not).

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -129,22 +122,36 @@ public AcrClientCredentials(string aadAccessToken, string loginUrl, string tenan
+                    acrClient.LoginUri = this._loginUrl;
+                }
+                // if the login uris don't match
+                else if (!acrClient.LoginUri.ToLower().Contains(this._loginUrl.ToLower()))
```
Should this be `ToLower` or `ToLowerInvariant` ?

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs:
```diff
@@ -153,40 +160,55 @@ public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, C
+        private static string ProcessLoginUrl(string loginUrl)
+        {
+            // Proofing in case passed in loginurl includes https start.
+            if (loginUrl.ToLower().StartsWith("https://"))
```
Does it make sense to add a `Debug.Assert(!loginUrl.StartsWith("http://"))` to make sure you never use non https?

AlexGhiondea reviewed (changes requested):
Great changes!!! I had a couple of small questions left!

Commit:
Tosin: Address PR comments / minor changes.

Tosin: rename classes and modify comments.

Commit:
Tosin: Add GetBlobOAuth Test

Make minor changes to tests. Add GetBlobAAD test.

Tosin: Address pr comments. Other style fixes.

Commit:
Tosin: change use of StringComparer.InvariantCultureIgnoreCase to StringComparer.OrdinalIgnoreCase as it is not available in .netstandard 1.4 a dependency of this project.

AlexGhiondea commented on sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs:
```diff
@@ -9,16 +9,16 @@
+    /// Simple authentication class for use in clients used by Token classes.
     /// </summary>
-    public class TokenCredentials : ServiceClientCredentials
+    internal class TokenCredentials : ServiceClientCredentials
```
Was this type public before? If it was, this is a breaking change.

adewaleo replied:
Yes it was Public earlier in this PR, however, this class is being introduced for the first time in this PR.

Commit:
Tosin: update package version from 0.9.0 to 0.10.0 and release notes.

Skip tests that should be run live.
